metrics.nim

```diff
@@ -324,22 +341,28 @@ else:
 
 proc incCounter(counter: Counter, amount: int64|float64 = 1, labelValues: LabelsParam = @[]) =
   when defined(metrics):
-    var timestamp = getTime().toMilliseconds()
+    try:
+      var timestamp = getTime().toMilliseconds()
 
-    if amount < 0:
-      raise newException(ValueError, "Counter.inc() cannot be used with negative amounts.")
+      if amount < 0:
+        raise newException(ValueError, "Counter.inc() cannot be used with negative amounts.")
 
-    let labelValuesCopy = validateCounterLabelValues(counter, labelValues)
+      let labelValuesCopy = validateCounterLabelValues(counter, labelValues)
 
-    atomicAdd(counter.metrics[labelValuesCopy][0].value.addr, amount.float64)
-    atomicStore(cast[ptr int64](counter.metrics[labelValuesCopy][0].timestamp.addr), timestamp.addr, ATOMIC_SEQ_CST)
+      atomicAdd(counter.metrics[labelValuesCopy][0].value.addr, amount.float64)
+      atomicStore(cast[ptr int64](counter.metrics[labelValuesCopy][0].timestamp.addr), timestamp.addr, ATOMIC_SEQ_CST)
 
-    pushMetrics(name = counter.name,
-                value = counter.metrics[labelValuesCopy][0].value,
-                increment = amount.float64,
-                metricType = "c",
-                timestamp = timestamp,
-                sampleRate = counter.sampleRate)
+      pushMetrics(name = counter.name,
+                  value = counter.metrics[labelValuesCopy][0].value,
+                  increment = amount.float64,
+                  metricType = "c",
+                  timestamp = timestamp,
+                  sampleRate = counter.sampleRate)
+    except Exception as e:
+      try:
+        write stderr, "metrics: " & e.msg
```

Comment:
Why would you let erroneous API usage slide with just a stderr message that no one will see?

Reply:
because `metrics`, just like logging, are non-critical and should not be intrusive for the caller

Reply:
It doesn't matter if they're critical or not. If the caller makes mistakes, the caller gets exceptions.

Reply:
well, that misses the point a bit: there are more exceptions coming out of here than the caller reasonably can handle, even when they haven't made a mistake per se - they just wanted to increase a counter  but for whatever reason, the metrics library can't deal with it. there's nothing meaningful the caller can do at that point so it's a feature of the library that it has a strategy handling the errors for the caller.

Reply:
Another way to look at this: if it's ever needed, the `inc` function is not the right place to be reporting errors - instead, at library setup time, one could imagine registering an error handling callback that would do more sophisticated error handling.

Reply:
yet another way to look at it: I'm increasing my counter, but the user unplugged the network card and now the UDP socket is dead - should the application crash? should I, the user of the library, have to deal with this in my eth2 state transition function?

Reply:
No, the `exportChan.trySend()` in `pushMetrics` should not crash and, at the same time, any error obfuscation should not be done in `incCounter()`.

Reply:
well, in the current implementation, that's the exception tracking effect it has, thus that's the API that metrics exposes to the user is the following: _any_ `Exception` may happen when you call this API

Reply:
Jacek's point is that the library should offer an API for manipulating the counters that offers the illusion that the operation cannot fail. The internal code of the library should strive to bring this illusion as closer to reality as possible. If a failure still happens, it can be reported with a global error hook function installed by the application. Such a function is likely to log the error for example.

All of this simplifies a single goal for the user - guaranteeing that all errors of interest are handled. Since a failure to manipulate a counter is very rarely an "error of interest", leaving it as a possibility in the API makes the life of the user harder.

You can still have another error-raising API potentially that some users might be interested in using.